instrumentation/log4j/log4j-2.13.2/library/src/main/java/io/opentelemetry/instrumentation/log4j/v2_13_2/LogEventMapper.java

```diff
@@ -0,0 +1,141 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.instrumentation.log4j.v2_13_2;
+
+import static io.opentelemetry.instrumentation.api.log.LoggingContextConstants.SPAN_ID;
+import static io.opentelemetry.instrumentation.api.log.LoggingContextConstants.TRACE_FLAGS;
+import static io.opentelemetry.instrumentation.api.log.LoggingContextConstants.TRACE_ID;
+
+import io.opentelemetry.api.common.AttributeKey;
+import io.opentelemetry.api.common.Attributes;
+import io.opentelemetry.api.common.AttributesBuilder;
+import io.opentelemetry.api.trace.Span;
+import io.opentelemetry.api.trace.SpanContext;
+import io.opentelemetry.api.trace.SpanId;
+import io.opentelemetry.api.trace.TraceFlags;
+import io.opentelemetry.api.trace.TraceId;
+import io.opentelemetry.api.trace.TraceState;
+import io.opentelemetry.context.Context;
+import io.opentelemetry.sdk.logs.LogBuilder;
+import io.opentelemetry.sdk.logs.data.Severity;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import org.apache.logging.log4j.Level;
+import org.apache.logging.log4j.Marker;
+import org.apache.logging.log4j.core.LogEvent;
+import org.apache.logging.log4j.message.Message;
+import org.apache.logging.log4j.util.ReadOnlyStringMap;
+
+final class LogEventMapper {
+
+  // Visible for testing
+  static final AttributeKey<String> ATTR_LOGGER_NAME = AttributeKey.stringKey("logger.name");
+  static final AttributeKey<String> ATTR_THREAD_NAME = AttributeKey.stringKey("thread.name");
+  static final AttributeKey<Long> ATTR_THREAD_ID = AttributeKey.longKey("thread.id");
+  static final AttributeKey<Long> ATTR_THREAD_PRIORITY = AttributeKey.longKey("thread.priority");
+  static final AttributeKey<String> ATTR_THROWABLE_NAME = AttributeKey.stringKey("throwable.name");
+  static final AttributeKey<String> ATTR_THROWABLE_MESSAGE =
+      AttributeKey.stringKey("throwable.message");
+  static final AttributeKey<List<String>> ATTR_NDC = AttributeKey.stringArrayKey("ndc");
+  static final AttributeKey<String> ATTR_FQCN = AttributeKey.stringKey("fqcn");
+  static final AttributeKey<String> ATTR_MARKER = AttributeKey.stringKey("marker");
+
+  private static final Map<Level, Severity> LEVEL_SEVERITY_MAP;
+
+  static {
+    Map<Level, Severity> levelSeverityMap = new HashMap<>();
+    levelSeverityMap.put(Level.ALL, Severity.TRACE);
```

Comment:
Use a method with a `switch` instead of a Map - complete enum switches get compiled into a jump table based on the enum ordinal, which is highly efficient. Errorprone verifying the switch is complete is added icing.

Alternative is `EnumMap`, but we'd only prefer it for an incomplete mapping which doesn't become a jump table.

Reply:
Wow I didn't know that about enum switch compilation. Cool!